keps/sig-node/0014-runtime-class.md

```diff
@@ -0,0 +1,359 @@
+that are generally static for the RuntimeClass, but dynamically interpreted by the runtime. They
+provide a way of specifying options that are tightly coupled with the class of runtime, but should
+not be used to enumerate general pod options as this would lead to a combinatorial explosion of
+RuntimeClass definitions.
+
+Examples of runtime parameters include:
+
+```
+// Decide which underlying OCI runtime should be used. This replaces the
+// annotations used by CRI-o, containerd & frakti to select between sandboxed
+// runtimes.
+"oci-runtime": "runsc"
+
+// Change the default set of capabilities used by the runtime. This is a good
+// way of surfacing behaviors that were previously implicit in docker but not
+// well defined in Kubernetes.
+"default-capabilities": "[]"
+
+// Set the infrastructure container image. More generally, configuring options
+// not otherwise exposed in the Kubrenetes API.
+"infra-image": "example.com/self-monitoring-pod:v2"
+```
+
+_Note: RuntimeClass is intended to be consumed by a single runtime implementation, so parameter keys
+do not need to be "namespaced"._
+
+In contrast, the following are examples of parameters that are better left to annotations or other
+extension mechanisms:
+
+```
+// Configuring the smack LSM for a runtime that adds support. This is an example
+// of a feature that will typically have multiple valid configurations, and is
+// better left to a pod-scoped mechanism like annotations.
+"smack-profile": "foo-bar"
+
+// Propagate Kubelet version information to the runtime. Tight coupling of a
+// RuntimeClass to nodes should generally be avoided. This will create churn
+// during upgrades.
+"kubelet-version": "1.2.3"
+```
+
+Minimal validation is done on parameters (copied from StorageClass):
+
+1. Keys must not be empty.
+2. A maximum of 512 parameters may be specified.
+3. Total size of keys+values must be less than 256K
+
+The runtime parameters are passed to the CRI as part of the `RunPodSandboxRequest`:
+
+```proto
+message RunPodSandboxRequest {
+    // Configuration for creating a PodSandbox.
+    PodSandboxConfig config = 1;
+    // Configuration for the runtime.
+    map<string, string> RuntimeParameters = 2;
+}
+```
+
+### Versioning, Updates, and Rollouts
+
+Getting upgrades and rollouts right is a very nuanced and complicated problem. For the initial alpha
+implementation, we will kick the can down the road by making the `RuntimeClassSpec` **immutable**,
+thereby requiring changes to be pushed as a newly named RuntimeClass instance. This means that pods
+must be updated to reference the new RuntimeClass, and comes with the advantage of native support
+for rolling updates through the same mechanisms as any other application update.
+
+This conservative approach is preferred since it's much easier to relax constraints in a backwards
+compatible way than tighten them. We should revisit this decision prior to graduating RuntimeClass
+to beta.
```

Comment:
Are you thinking in beta to be able to update a given RuntimeClass params, and that kicks the pod rolling update automatically without updating the class name ?